### PR TITLE
[Transformer] Use float16 input and output for softmax in mixed-precision training

### PR DESCRIPTION
**Pull request by zongweiz**
Tensorflow softmax GPU internally internally performs math operations using float32 for numeric stability. So in mixed-precision training, we can keep the input and output of softmax in float16 for better performance (without manual casting and also save memory bandwidth).

### CODE REVIEW

**Comment by zongweiz** [comment]
softmax GPU kernel code:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/softmax_op_gpu.cu.cc#L72
accumulator_type is forced to float when input type is Eigen::half.

**Comment by guptapriya** [comment]
LGTM, will let Reed approve as he has more context.

**Comment by reedwm** [comment]
I think the added comment might not be entirely correct, as I think Softmax still does some operations in float16 when it's inputs/outputs are in float16. We will have to take a closer look.

In any case, I think we should try doing Softmax entirely in float32 before submitting this PR, since otherwise this might cause convergence issues. And we have to check XLA as well.

/CC @tfboyd if you have any thoughts

**Comment by zongweiz** [comment]
I looked at the SoftmaxOpGPU op kernel, it seems that only input, max, output logits are in f16, other intermediate values and computation are in f32. @reedwm do you mind double check? Thanks. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/softmax_op_gpu.cu.cc#L139

**Comment by reedwm** [comment]
@zongweiz you are correct.
